### ISSUE TITLE
Not possible to use json_encoders with NamedTuple

### ISSUE THREAD

**Issue by barisione** [pydantic/pydantic#3009]
### Checks

* [x] I added a descriptive title to this issue
* [x] I have searched (google, github) for similar issues and couldn't find anything
* [x] I have read and followed [the docs](https://pydantic-docs.helpmanual.io/) and still think this is a bug

# Bug

Output of `python -c "import pydantic.utils; print(pydantic.utils.version_info())"`:
```
$ python3 -c "import pydantic.utils; print(pydantic.utils.version_info())"
             pydantic version: 1.8.2
            pydantic compiled: True
                 install path: /usr/local/lib/python3.8/site-packages/pydantic
               python version: 3.8.5 (default, Jul 21 2020, 10:42:08)  [Clang 11.0.0 (clang-1100.0.33.17)]
                     platform: macOS-10.16-x86_64-i386-64bit
     optional deps. installed: ['typing-extensions']
```

It's possible to define a custom encoder for classes defined by the user, including data classes, but not base types, like tuples. I do understand this behaviour but it also applies to named tuples:

```py
from dataclasses import dataclass
from typing import NamedTuple, Tuple

from pydantic import BaseModel


@dataclass
class MyDataClass:
    value: str


class MyNamedTuple(NamedTuple):
    value: str


class MyModel(BaseModel):
    dc: MyDataClass
    nt: MyNamedTuple
    l: Tuple[str, ...]

    class Config:
        arbitrary_types_allowed = True
        json_encoders = {
            MyDataClass: lambda v: f"Encoded dataclass: {v.value}",
            MyNamedTuple: lambda v: f"Encoded NamedTuple: {v.value}",
            list: lambda v: f"Encoded list: {v}",
        }


model = MyModel(
    dc=MyDataClass("my dataclass"),
    nt=MyNamedTuple("my NamedTuple"),
    l=("my", "normal", "tuple"),
)
print(model.json(indent=2))

```

I would expect to get this:
```
{
  "dc": "Encoded dataclass: my dataclass",
  "nt": "Encoded NamedTuple: my NamedTuple,
  "l": [
    "my",
    "normal",
    "tuple"
  ]
}
```
But I get this:
```
{
  "dc": "Encoded dataclass: my dataclass",
  "nt": [
    "my NamedTuple"
  ],
  "l": [
    "my",
    "normal",
    "tuple"
  ]
}
```

**Comment by PrettyWood** [comment]
Hi @barisione 
You can't customize `json_encoder` for builtin and supported types.
It's the same for your `list` (`tuple` in fact), the custom serializer is not even triggered.
This may be fixed by #2745.
There will be a second afterwards: `MyNamedTuple` custom serializer will not be triggered by the fallback nor by the "builtin" wrapper :/

**Comment by dmontagu** [comment]
This can be resolved now through the use of `@field_serializer` or serializer annotations. Here is one way using the `PlainSerializer` annotation:

```python
from datetime import date

from typing_extensions import Annotated

import pydantic
from pydantic.annotated_arguments import PlainSerializer


from dataclasses import dataclass
from typing import NamedTuple, Tuple

from pydantic import BaseModel


@dataclass
class MyDataClass:
    value: str


class MyNamedTuple(NamedTuple):
    value: str


class MyModel(BaseModel):
    dc: Annotated[MyDataClass, PlainSerializer(lambda v: f"Encoded dataclass: {v.value}")]
    nt: Annotated[MyNamedTuple, PlainSerializer(lambda v: f"Encoded NamedTuple: {v.value}")]
    l: Tuple[str, ...]


model = MyModel(
    dc=MyDataClass("my dataclass"),
    nt=MyNamedTuple("my NamedTuple"),
    l=("my", "normal", "tuple"),
)
print(model.model_dump_json(indent=2))
"""
{
  "dc": "Encoded dataclass: my dataclass",
  "nt": "Encoded NamedTuple: my NamedTuple",
  "l": [
    "my",
    "normal",
    "tuple"
  ]
}
"""
```

`@field_serializer` is a decorator like `@validator` (replaced by `@field_validator` in v2), and has a slightly different syntax but similar functionality.
